File: app/backend/routers/projects/services.py
import json, asyncio
from fastapi import WebSocket
from database import database
from .helpers.service_invoker import handle_db_command, DBoperations, FileOperations, GeneralOperations, handle_file_command, push_schema, handle_package_command, handle_general_commands
from .helpers.cmdhandlers import handle_shell_command, handle_cd_command
from .helpers.containerhelper import check_service_health, check_container_health, check_service_exists
from helpers.servicestates import send_service_status, services_alive
from helpers.structlogger import logger
from helpers.queries.servicequery import get_service_query


def get_project_services(project_id: str, websocket: WebSocket = None) -> dict:
    if project_id not in services_alive:
        services_alive[project_id] = {"container": True}
    if websocket:
        services_alive[project_id]["ws"] = websocket
    return services_alive[project_id]

async def start_service(container, project_id: str, project_name: str, service: str, q: asyncio.Queue):
    project_services = get_project_services(project_id)

    try:
        container.reload()
        if container.status != "running":
            await q.put(f"[✗] Container is not running\n")
            project_services.update({k: {"enabled": False, "status": "dead"} for k in project_services if k != "container"})
            project_services["container"] = False
            await send_service_status(q, project_services)
            return
    except Exception as e:
        await q.put(f"[✗] Container unreachable: {str(e)}\n")
        services_alive.pop(project_id, None)
        await send_service_status(q, {"container": False})
        return

    row = await database.fetch_one(
        get_service_query,
        {"project_id": project_id, "category": service},
    )
    print(f"service: {service}, row: {row}")

    frontend_root = row["frontend_root"] or f"/app/workspace/{project_name}"
    backend_root = row["backend_root"] or f"/app/workspace/{project_name}"

    if not row:
        await q.put(f"[✗] No {service} service configured\n")
        project_services[service] = {"enabled": False, "port": 0, "status": "missing", "name": service}
        await send_service_status(q, project_services)
        return

    command = row["custom_start_command"] or row["default_start_command"]
    port = row["default_port"] or 0

    if service == "backend":
        container.exec_run(["sh", "-c", f"cd {backend_root} && {command} >/tmp/{service}.log 2>&1 &"], detach=True)
    elif service == "frontend":
        container.exec_run(["sh", "-c", f"cd {frontend_root} && {command} >/tmp/{service}.log 2>&1 &"], detach=True)
    elif service == "database":
        container.exec_run(["sh", "-c", f"{command} >/tmp/{service}.log 2>&1"], detach=True)
        await asyncio.sleep(0.2)
        check_result = container.exec_run(f'su - postgres -c "psql -lqt | cut -d \\| -f 1 | grep -qw myapp && echo exists || echo missing"')
        if b"exists" in check_result.output:
            await push_schema(container, project_id, q)
        else:
            await q.put(f"[ℹ] Database 'myapp' not found\n")

    print(f"executed command for {service}: {command}")
    await q.put(f"[→] Starting {row['name']} ({service})...\n")
    await asyncio.sleep(0.2)
    is_running = await check_service_health(container, service)

    if is_running:
        await q.put(f"[✓] {row['name']} is running on port {port}\n")
        project_services[service] = {"enabled": True, "port": port, "status": "running", "name": row['name'], "category": row['category']}
        await send_service_status(q, project_services)
    else:
        log_file = f'/tmp/{service}.log'
        log_check = container.exec_run(f"tail -20 {log_file}")
        log_output = log_check.output.decode() if log_check.output else "No logs available"
        await q.put(f"[!] Failed to start {row['name']}. Logs:\n{log_output}\n")
        project_services[service] = {"enabled": False, "port": port, "status": "failed", "name": row['name'], "logs": log_output}
        await send_service_status(q, project_services)


def is_log_event(line: str) -> bool:
    try:
        data = json.loads(line)
        return data.get("type") == "LOG_EVENT"
    except (json.JSONDecodeError, AttributeError):
        return False

# Command handlers
async def handle_json_command(container, payload: dict, current_dir: str, q: asyncio.Queue, project_id: str, project_name: str):
    if payload.get('type') == 'START_SERVICE':
        await start_service(container, project_id, project_name, payload.get('service'), q)
        return "", current_dir

    if payload.get('operation') in DBoperations:
        await handle_db_command(container, payload, q, project_id)
        return "", current_dir

    if payload.get('type') in FileOperations:
        await handle_file_command(container, payload, q)
        return "", current_dir

    if payload.get('type') == 'PACKAGE':
        await handle_package_command(container, payload, q)
        return None, current_dir
    
    if payload.get('type') in GeneralOperations:
        await handle_general_commands(container, payload, q)
        return "", current_dir

    return "Command handled\n", current_dir


async def process_command(container, cmd: str, current_dir: str, q: asyncio.Queue, project_id: str, project_name: str):
    cmd = cmd.strip()
    if not cmd:
        return "", current_dir

    if cmd.startswith("cd "):
        return handle_cd_command(cmd, current_dir)

    if cmd.startswith("{") and cmd.endswith("}"):
        try:
            payload = json.loads(cmd)
            logger.info(f"Received JSON payload: {payload}")
            result = await handle_json_command(container, payload, current_dir, q, project_id, project_name)
            await database.execute(
                "UPDATE projects SET last_online = NOW() WHERE project_id = :project_id",
                {"project_id": project_id}
            )
            return result
        except Exception as e:
            print(f"Error handling JSON command: {e}")
            return f"Error handling command: {str(e)}\n", current_dir

    result = handle_shell_command(container, cmd, current_dir)
    await database.execute(
        "UPDATE projects SET last_online = NOW() WHERE project_id = :project_id",
        {"project_id": project_id}
    )
    return result



async def tail_logd(container, send_queue: asyncio.Queue):
    try:
        exec_result = container.exec_run(
            "tail -f /var/log/logd.log",
            stream=True,
            tty=False,
            detach=False,
            socket=False
        )
        loop = asyncio.get_event_loop()
        read_queue = asyncio.Queue()

        def read_stream():
            for chunk in exec_result.output:
                lines = chunk.decode("utf-8", errors="replace").splitlines()
                for line in lines:
                    line = line.strip()
                    if line and is_log_event(line):
                        loop.call_soon_threadsafe(read_queue.put_nowait, line)

        loop.run_in_executor(None, read_stream)

        while True:
            try:
                line = await asyncio.wait_for(read_queue.get(), timeout=1.0)
                print(f"Forwarding log event: {line}")
                await send_queue.put(line)
            except asyncio.TimeoutError:
                continue

    except Exception as e:
        logger.warning(f"logd tail stopped: {e}")

File: backend/handlers/projects/services.go
package handlers

import (
	"context"
	"database/sql"
	"fmt"
	"strings"
	"time"
)

type projectServiceRow struct {
	Name, Category string
	DefaultCommand sql.NullString
	DefaultPort    sql.NullInt64
	CustomCommand  sql.NullString
	FrontendRoot   sql.NullString
	BackendRoot    sql.NullString
}

func startProjectService(ctx context.Context, db *sql.DB, helper *ContainerHelper, containerID, projectID, projectName, service string, send chan<- string) error {
	inspect, err := helper.Docker.ContainerInspect(ctx, containerID)
	if err != nil {
		queueProjectMessage(ctx, send, "[✗] Container unreachable: "+err.Error()+"\n")
		updateProjectServices(ctx, projectID, send, func(services map[string]any) {
			for key := range services {
				if key != "container" {
					delete(services, key)
				}
			}
			services["container"] = false
		})
		return nil
	}
	if inspect.State == nil || !inspect.State.Running {
		queueProjectMessage(ctx, send, "[✗] Container is not running\n")
		updateProjectServices(ctx, projectID, send, func(services map[string]any) {
			for key := range services {
				if key != "container" {
					services[key] = map[string]any{"enabled": false, "status": "dead"}
				}
			}
			services["container"] = false
		})
		return nil
	}

	var row projectServiceRow
	err = db.QueryRowContext(ctx, `
		SELECT s.name, s.category, s.default_start_command, s.default_port, ps.custom_start_command, p.frontend_root, p.backend_root
		FROM project_services ps
		JOIN services s ON s.id = ps.service_id
		JOIN projects p ON p.project_id = ps.project_id
		WHERE ps.project_id = $1 AND s.category = $2
	`, projectID, service).Scan(&row.Name, &row.Category, &row.DefaultCommand, &row.DefaultPort, &row.CustomCommand, &row.FrontendRoot, &row.BackendRoot)
	if err == sql.ErrNoRows {
		queueProjectMessage(ctx, send, "[✗] No "+service+" service configured\n")
		updateProjectServices(ctx, projectID, send, func(services map[string]any) {
			services[service] = map[string]any{"enabled": false, "port": 0, "status": "missing", "name": service}
		})
		return nil
	}
	if err != nil {
		return err
	}

	command := row.DefaultCommand.String
	if row.CustomCommand.Valid && row.CustomCommand.String != "" {
		command = row.CustomCommand.String
	}
	frontendRoot, backendRoot := row.FrontendRoot.String, row.BackendRoot.String
	if frontendRoot == "" {
		frontendRoot = "/app/workspace/" + projectName
	}
	if backendRoot == "" {
		backendRoot = "/app/workspace/" + projectName
	}
	switch service {
	case "frontend":
		err = helper.ExecDetached(ctx, containerID, "cd "+shellQuote(frontendRoot)+" && "+command+" >/tmp/frontend.log 2>&1 &")
	case "backend":
		err = helper.ExecDetached(ctx, containerID, "cd "+shellQuote(backendRoot)+" && "+command+" >/tmp/backend.log 2>&1 &")
	case "database":
		err = helper.ExecDetached(ctx, containerID, command+" >/tmp/database.log 2>&1")
	}
	if err != nil {
		return err
	}
	if service == "database" {
		time.Sleep(200 * time.Millisecond)
		output, _, checkErr := runContainerCommand(ctx, helper, containerID, "", []string{"sh", "-c", `su - postgres -c "psql -lqt | cut -d \| -f 1 | grep -qw myapp && echo exists || echo missing"`})
		if checkErr == nil && strings.Contains(output, "exists") {
			if err := pushProjectSchema(ctx, db, helper, containerID, projectID, send); err != nil {
				return err
			}
		} else {
			queueProjectMessage(ctx, send, "[ℹ] Database 'myapp' not found\n")
		}
	}

	queueProjectMessage(ctx, send, "[→] Starting "+row.Name+" ("+service+")...\n")
	time.Sleep(200 * time.Millisecond)
	port := int(row.DefaultPort.Int64)
	if projectServiceHealthy(ctx, helper, containerID, service) {
		queueProjectMessage(ctx, send, fmt.Sprintf("[✓] %s is running on port %d\n", row.Name, port))
		updateProjectServices(ctx, projectID, send, func(services map[string]any) {
			services[service] = map[string]any{"enabled": true, "port": port, "status": "running", "name": row.Name, "category": row.Category}
		})
		return nil
	}
	logOutput, _, logErr := runContainerCommand(ctx, helper, containerID, "", []string{"tail", "-20", "/tmp/" + service + ".log"})
	if logErr != nil || logOutput == "" {
		logOutput = "No logs available"
	}
	queueProjectMessage(ctx, send, "[!] Failed to start "+row.Name+". Logs:\n"+logOutput+"\n")
	updateProjectServices(ctx, projectID, send, func(services map[string]any) {
		services[service] = map[string]any{"enabled": false, "port": port, "status": "failed", "name": row.Name, "logs": logOutput}
	})
	return nil
}

func projectServiceHealthy(ctx context.Context, helper *ContainerHelper, containerID, service string) bool {
	// These category ports are part of the Python terminal protocol and intentionally ignore service-row overrides.
	port := map[string]string{"frontend": "5173", "backend": "8000", "database": "5432"}[service]
	if port == "" {
		return false
	}
	for attempt := 0; attempt < 10; attempt++ {
		output, exitCode, err := runContainerCommand(ctx, helper, containerID, "", []string{"sh", "-c", "netstat -tuln | grep " + port})
		if err == nil && exitCode == 0 && output != "" {
			return true
		}
		select {
		case <-ctx.Done():
			return false
		case <-time.After(500 * time.Millisecond):
		}
	}
	return false
}
